### ISSUE TITLE
simsce_supervised 代码实现问题

### ISSUE THREAD

**Issue by rookiebird** [zhoujx4/NLP-Series-sentence-embeddings#1]
从知乎来的，参考大佬的代码实现了simsce_supervised，loss 函数计算的的时候

def compute_loss(y_pred,lamda=0.05):
    row = torch.arange(0,y_pred.shape[0],3,device='cuda')
    col = torch.arange(y_pred.shape[0],device='cuda')
    col = torch.where(col % 3 != 0)[0].cuda()
    # 每个origin 对应的 entail 的位置
    y_true = torch.arange(0,len(col),2,device='cuda')
    similarities = F.cosine_similarity(y_pred.unsqueeze(1), y_pred.unsqueeze(0), dim=2) 
    sim = sim - torch.eye(y_pred.shape[0], device=device) * 1e12
    #丢弃对角矩阵即自身相等的loss
    similarities = torch.index_select(similarities,0,row)
    similarities = torch.index_select(similarities,1,col)
    #除以温度超参数
    loss = F.cross_entropy(similarities,y_true)
    return torch.mean(loss)

torch.index_select(similarities,1,col) 丢弃了所有列 index%3==0的值感觉是不太对的，应该丢弃对角线上的值

**Comment by zhoujx4** [comment]
没有错的，index%3==0是这个句子自己与自己， 应该把它抛弃掉，你再理解一下




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zhoujx4/NLP-Series-sentence-embeddings"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年12月6日(星期一) 上午10:21
***@***.***&gt;;
***@***.***&gt;;
主题:&nbsp;[zhoujx4/NLP-Series-sentence-embeddings] simsce_supervised 代码实现问题 (Issue #1)





 
从知乎来的，参考大佬的代码实现了simsce_supervised，loss 函数计算的的时候
 
def compute_loss(y_pred,lamda=0.05):
 row = torch.arange(0,y_pred.shape[0],3,device='cuda')
 col = torch.arange(y_pred.shape[0],device='cuda')
 col = torch.where(col % 3 != 0)[0].cuda()
 # 每个origin 对应的 entail 的位置
 y_true = torch.arange(0,len(col),2,device='cuda')
 similarities = F.cosine_similarity(y_pred.unsqueeze(1), y_pred.unsqueeze(0), dim=2)
 sim = sim - torch.eye(y_pred.shape[0], device=device) * 1e12 #丢弃对角矩阵即自身相等的loss similarities = torch.index_select(similarities,0,row) similarities = torch.index_select(similarities,1,col) #除以温度超参数 loss = F.cross_entropy(similarities,y_true) return torch.mean(loss)  
torch.index_select(similarities,1,col) 丢弃了所有列 index%3==0的值感觉是不太对的，应该丢弃对角线上的值
 
—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by rookiebird** [comment]
我指的是  `similarities = torch.index_select(similarities,1,col)` 这一句，因为之前 `col = torch.where(col % 3 != 0)[0].cuda()` ， 假设batch_size为9，那么对于第一个样本，只需要丢弃第similarity 矩阵第一行index为 0 的元素，但是`similarities = torch.index_select(similarities,1,col)`  这句还丢弃了第一行index 为3 和 6的元素
